### PR TITLE
storage: handle unordered key indices during upsert

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

We were incorrectly rehydrating rows when the key indices weren't sorted. Since the intermediate row produced by `thin` is just an implementation detail that is never externally observed and `rehydrate` is its inverse function it's safe to always sort the indices in both cases.

I changed the testcase in a way that was failing on `main` and is passing with this change.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### CODE REVIEW

**Review by umanwizard — COMMENTED** [review]
Uhhhhhhhhhh I'm really not sure why I thought I needed a bunch of gobs of code to solve this. Thanks.

**Review by umanwizard — CHANGES_REQUESTED** [review]
actually, I think this is wrong, please wait before merging (confirming now)

File: src/storage/src/render/upsert.rs [hunk 1]
```diff
@@ -602,7 +606,7 @@ mod tests {
     fn test_rehydrate_thin_multiple() {
         let mut packer = Row::default();
 
-        let key_indices = vec![2, 4];
+        let key_indices = vec![4, 2];
         let key = Row::pack([Datum::String("key1"), Datum::String("key2")]);
```
Comment (umanwizard) [inline src/storage/src/render/upsert.rs:610]: This should now be `Row::pack([Datum::String("key2"), Datum::String("key1")]);`, but then the test will fail. So I think the logic in the PR is actually wrong

Comment (umanwizard) [reply src/storage/src/render/upsert.rs:610]: The problem is that in `rehydrate`, you won't know where to put the key columns back into the values. You just insert them in order, which is wrong.

I guess this is why I made the `BTreeMap` in https://github.com/MaterializeInc/materialize/pull/12687 -- you do need to track both the key column index and the value column index for each of the thinned columns.

**Comment by petrosagg** [comment]
Turns out this is incorrect as @umanwizard pointed out. Closing

**Pull request closed without merging**